fcm_django/admin.py

```diff
@@ -1,79 +1,104 @@
+        if total_failure == 0:
+            return
+        if is_topic:
+            message = ngettext_lazy(
+                "A device failed to un/subscribe to topic. %(count)d device was "
+                "marked as inactive.",
+                "Some devices failed to un/subscribe to topic. %(count)d devices "
+                "were marked as inactive.",
+                total_failure,
+            )
         else:
-            return ("name", "device_id")
+            message = ngettext_lazy(
+                "A message failed to send. %(count)d device was marked as " "inactive.",
+                "Some messages failed to send. %(count)d devices were marked as "
+                "inactive.",
+                total_failure,
+            )
+        self.message_user(
+            request,
+            message % {"count": total_failure},
+            level=messages.WARNING,
+        )
 
-    def send_messages(self, request, queryset, bulk=False, data=False):
+    def send_messages(self, request, queryset, bulk=False):
         """
         Provides error handling for DeviceAdmin send_message and
         send_bulk_message methods.
         """
-        ret = []
-        errors = []
         total_failure = 0
 
         for device in queryset:
             if bulk:
-                if data:
-                    response = queryset.send_message(
-                        data={"Nick": "Mario"}
-                    )
-                else:
-                    response = queryset.send_message(
-                        title="Test notification",
-                        body="Test bulk notification"
+                response = queryset.send_message(
+                    Message(
+                        notification=Notification(
+                            title="Test notification", body="Test bulk notification"
+                        )
                     )
+                )
+                total_failure = len(response["deactivated_registration_ids"])
+                break
             else:
-                if data:
-                    response = device.send_message(data={"Nick": "Mario"})
-                else:
-                    response = device.send_message(
-                        title="Test notification",
-                        body="Test single notification"
+                response = device.send_message(
+                    Message(
+                        notification=Notification(
+                            title="Test notification", body="Test single notification"
+                        )
                     )
-            if response:
-                ret.append(response)
-
-            failure = int(response['failure'])
-            total_failure += failure
-            errors.append(str(response))
+                )
+                if type(response) != SendResponse:
+                    total_failure += 1
 
-            if bulk:
-                break
-
-        if ret:
```

Comment:
This should not be removed, as it showed the actual FCM errors to the user. Showing merely "failed to send" is fairly useless

Reply:
+1